omnichain/swap/contracts/Swap.sol

```diff
@@ -47,26 +49,30 @@ contract Swap is zContract {
             recipientAddress = recipient;
         }
 
-        uint256 outputAmount = SwapHelperLib._doSwap(
+        (address gasZRC20, uint256 gasFee) = IZRC20(targetTokenAddress)
+            .withdrawGasFee();
+
+        inputForGas = SwapHelperLib.swapTokensForExactTokens(
+            systemContract.wZetaContractAddress(),
+            systemContract.uniswapv2FactoryAddress(),
+            systemContract.uniswapv2Router02Address(),
+            zrc20,
+            gasFee,
+            gasZRC20,
+            amount
+        );
+
+        outputAmount = SwapHelperLib._doSwap(
             systemContract.wZetaContractAddress(),
             systemContract.uniswapv2FactoryAddress(),
             systemContract.uniswapv2Router02Address(),
             zrc20,
-            amount,
+            amount - inputForGas,
             targetTokenAddress,
             0
         );
 
-        (address gasZRC20, uint256 gasFee) = IZRC20(targetTokenAddress)
-            .withdrawGasFee();
-
-        if (gasZRC20 != targetTokenAddress) revert WrongGasContract();
-        if (gasFee >= outputAmount) revert NotEnoughToPayGasFee();
-
-        IZRC20(targetTokenAddress).approve(targetTokenAddress, gasFee);
-        IZRC20(targetTokenAddress).withdraw(
-            recipientAddress,
-            outputAmount - gasFee
-        );
+        IZRC20(gasZRC20).approve(gasZRC20, gasFee);
+        // IZRC20(targetTokenAddress).withdraw(recipientAddress, outputAmount);
```

Comment:
why you don't need the abi.encode? will check if is the same without... IZRC20(targetTokenAddress).withdraw(abi.encodePacked(recipientAddress), outputAmount);